### PR TITLE
fix: retain v2 indexes in codegen

### PR DESCRIPTION
**Pull request by edwardfoyle**
#### Description of changes

Maps new @index and @primaryKey directives back to their equivalent key representations to ensure the key attributes are not lost during the field pruning that happens when parsing connections.

The important chunks of logic are in `process-index.ts` and `process-primary-key.ts`. In these files, we look through all the fields on the model and if we find an `@index` or `@primaryKey`, respectively, we translate that into a "key" directive and attach it to the model. This allows other downstream codegen components to generate keys as they always have.

#### Description of how you validated changes
Added unit tests and verified the v2 codegen output of Android, iOS, JS, and Flutter are identical to the output of an equivalent v1 schema.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-codegen/blob/master/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by cjihrig — COMMENTED** [review]
LGTM

**Comment by edwardfoyle** [comment]
Hold off merging for a minute. Found an issue with Android codegen

File: packages/appsync-modelgen-plugin/src/utils/process-primary-key.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { CodeGenModel } from '../visitors/appsync-visitor';
+import { getDirective } from './fieldUtils';
+
+export const processPrimaryKey = (model: CodeGenModel) => {
+  const primaryKeyField = model.fields.find(field => getDirective(field)('primaryKey'));
```
Comment (phani-srikar) [inline packages/appsync-modelgen-plugin/src/utils/process-primary-key.ts:5]: Do you know if there is a validation check already in place to make sure we have only one `primaryKey` on a model?

File: packages/appsync-modelgen-plugin/src/visitors/appsync-visitor.ts [hunk 2]
```diff
@@ -293,6 +295,7 @@ export class AppSyncModelVisitor<
     };
   }
   processDirectives() {
+    this.processV2KeyDirectives();
```
Comment (phani-srikar) [inline packages/appsync-modelgen-plugin/src/visitors/appsync-visitor.ts:298]: Can we move this inside the `if (this.config.usePipelinedTransformer || this.config.transformerVersion === 2)` conditional below?

**Review by phani-srikar — CHANGES_REQUESTED** [review]
Looks Good. Minor comments.

File: packages/appsync-modelgen-plugin/src/utils/process-primary-key.ts (hunk 1 above)
Comment (cjihrig) [reply packages/appsync-modelgen-plugin/src/utils/process-primary-key.ts:5]: Yes, there is.

File: packages/appsync-modelgen-plugin/src/visitors/appsync-visitor.ts (hunk 2 above)
Comment (edwardfoyle) [reply packages/appsync-modelgen-plugin/src/visitors/appsync-visitor.ts:298]: sure. It's a noop if the new directives aren't present but I can move it inside the if statement

File: packages/appsync-modelgen-plugin/src/utils/process-primary-key.ts (hunk 1 above)
Comment (edwardfoyle) [reply packages/appsync-modelgen-plugin/src/utils/process-primary-key.ts:5]: not that I'm aware of but I can add a sanity check here

Comment (cjihrig) [reply packages/appsync-modelgen-plugin/src/utils/process-primary-key.ts:5]: @phani-srikar was your question about codegen or the `@primaryKey` directive? There is a [check](https://github.com/aws-amplify/amplify-cli/blob/master/packages/amplify-graphql-index-transformer/src/graphql-primary-key-transformer.ts#L125) and a regression test in the `@primaryKey` implementation.

Comment (phani-srikar) [reply packages/appsync-modelgen-plugin/src/utils/process-primary-key.ts:5]: I meant on the codegen side. Better to have some validation in place. Thanks Colin!

**Comment by edwardfoyle** [comment]
Updated to fix the Android issue and address other PR comments

**Review by phani-srikar — APPROVED** [review]
LGTM!

**Pull request merged**